### PR TITLE
[fixed] mechanisms

### LINKED ISSUE #1274: Motion Sensor and Obstructor can float
# Description

A Motion Sensor or Obstructor (in its background state) can float indefinitely, if you first place it, then place background behind it, then destroy the supporting tile. The background will fall down but the blob itself will stay afloat.
You can touch and build around the floating tile and it will never fall down. It can still be destroyed though.

https://i.imgur.com/4PoaNws.png

This issue is not fixed by PR #1243 .

# Reproduction Steps

- Go to Sandbox.
- Place a ladder.
- Place a Motion Sensor or Obstructor on top of the ladder.
- Place a wooden background on the Motion Sensor or Obstructor.
- Destroy the ladder.
- Notice the background falling down, but the blobs themselves staying afloat.

### LINKED ISSUE #1874: Obstructor hit by Keg will become bugged
## Description

When obstructor is hit by an exploding keg, it will cause obstructor to become bugged.

* The obstructor will not change its state between solid and non-solid anymore, even when power is supplied.
* Builder can build over the obstructor with new tiles.

It seems obstructor is not killed/despawned properly.

Happens easily in offline and online in Sandbox. Sometimes happens in TDM matches.

### LINKED ISSUE #1875: Some Components will leave blackness when destroyed
## Description

For example, spawn a bunch of component tiles, such as motion sensor.

Have it be destroyed by orb, bomb or keg.

https://i.imgur.com/ByFIFcK.png

It is not that much of a problem, but it seems like buggy behavior so I made this issue.

### LINKED ISSUE #2004: Components and devices have wrong frame when rejoining server
## Description

When building fresh

screen-24-04-10-16-51-04

After quitting and rejoining:

screen-24-04-10-16-51-28

* Lamp has light, but no light color
* Pressure Plate has wrong frame (not pushed down)
* Lever has wrong frame (not active)
* Spiker has an additional spike on top?
* Most wire components have the wrong frame (not active)

I guess `onReceiveCreateData()` could be used to set the necessary state on rejoining.

### LINKED ISSUE #2152: Spiker sometimes deals 2 hearts damage
## Description

Spiker triggered by motion sensor will sometimes deal 2 hearts of damage instead of 1.

Tested and verified in online Sandbox.

This is of significance in TDM since same traps exist there and one Spiker should not deal more than 1 damage per hit.

### LINKED ISSUE #2200: Coin Slot doesn't make coin sound
## Description

When entering 60 coins, it doesn't make a sound anymore. Both in online and offline.

### LINKED ISSUE #2201: Lever doesn't toggle sprite
## Description

When using a lever, it doesn't change its sprite, both in online and offline.

### PR DESCRIPTION
**Pull request by Gingerbeard5773**
## Status

- **READY**:

## Description

Mechanisms have been left behind with every update, causing various bugs.

* Fixes mechanism states & appearance not synchronizing to joining players on dedi.
* Fixes mechanism effects not appearing on client (lever & button).
* Fixes a problem where mechanism code was executed twice on localhost.
* Fixes a bug where players could not input items into the magazine due to their resolution.
* Improves the magazine so that it is easier to use.
* Fixed blood not appearing on the spiker.
* Improves the obstructor by giving it damage frames.
* Deletes two legacy files that become unnecessary with this PR. Don't scream at me for it because I genuinely believe it wont cause any problems for modders.

resolves #1274
resolves #1874
resolves #1875
resolves #2004
resolves #2152
resolves #2201
resolves #2200

Other PRs that are also closed by default due to this PR
#1879
#2007
#2202

Credits to @mugg91 , as some parts of this PR are from him.

### CODE REVIEW

**Review by Vam-Jam — CHANGES_REQUESTED** [review]
Dispensers can not be placed in sandbox or by the map loader. Results in TDM maps breaking

**Comment by Gingerbeard5773** [comment]
> Dispensers can not be placed in sandbox or by the map loader. Results in TDM maps breaking
resolved

File: Entities/Structures/Components/Load/Dispenser/Dispenser.as [hunk 1]
```diff
@@ -23,87 +22,22 @@ class Dispenser : Component
...
-		1.0f,                               // scale
-		3,                                  // ticks per frame
-		0.0f,                               // gravity
-		false);                             // self lit
-
 		if (isServer())
 		{
+			CMap@ map = getMap();
 			CBlob@[] blobs;
 			getMap().getBlobsAtPosition((offset * -1) * 8 + position, @blobs);
```
Comment (Vam-Jam) [inline Entities/Structures/Components/Load/Dispenser/Dispenser.as:29]: use map instead of getMap() here

**Review by Vam-Jam — DISMISSED** [review]
I played through a few TDM rounds, maps seem to be working correctly now, things worked when testing in sandbox. May need more testing but LGTM

**Review by Vam-Jam — APPROVED** [review]

**Review by asumagic — APPROVED** [review]
Also tested a bit manually in sandbox on dedi + tested TDM maps on dedi, saw nothing abnormal.

**Comment by mugg91** [comment]
Spiker seems to behave normally now, in both offline and online.
The only thing I found is that Lever will already by activated when placed; previously it was in deactivated state when placed.

**Comment by mugg91** [comment]
Just for future reference, this PR made it so that Obstructor and Sensor can't be built hanging from a row of blocks anymore.
It was more like a bug than a feature, so it is fine.

**Pull request merged**